sklearn/preprocessing/_discretization.py

```diff
@@ -316,4 +321,4 @@ def inverse_transform(self, Xt):
             bin_centers = (bin_edges[1:] + bin_edges[:-1]) * 0.5
             Xinv[:, jj] = bin_centers[np.int_(Xinv[:, jj])]
 
-        return Xinv
+        return Xinv.astype(self.dtype)
```

Comment:
When `if 'onehot' in self.encode` the  `_encoder` should take care of the dtype, so we don't need to convert it in that case.

We can use,
```
Xinv.astype(self.dtype, **_astype_copy_false(X))
```
with `_astype_copy_false` from `sklearn.fixes`.

Reply:
I don't have `astype` anymore in the second solution.